### PR TITLE
sessionctx,infoschema,executor,util: Show disk usage of a query in slow query and statement summary

### LINKED ISSUE #16883: Show disk usage of a query in slow query and statement summary
## Description

1. Show disk usage of a query in slow query log and SLOW_QUERY table. (https://pingcap.com/docs/stable/how-to/maintain/identify-abnormal-queries/identify-slow-queries/)
2. Show disk usage of a query in statement summary table (https://pingcap.com/docs/stable/reference/performance/statement-summary/#statement-summary-tables)
## Score

* 50

## Mentor(s)

* @SunRunAway 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* skills 1
* skills 1

## Learning Materials

* [Slow qeury]https://pingcap.com/docs/stable/how-to/maintain/identify-abnormal-queries/identify-slow-queries/
* [Statement summary]https://pingcap.com/docs/stable/reference/performance/statement-summary/#statement-summary-tables
* Chinese: [TiDB 精选技术讲解文章](https://github.com/pingcap/presentations/blob/master/hackathon-2019/reference-document-of-hackathon-2019.md)
* English: [Awesome-Database-Learning](https://github.com/pingcap/awesome-database-learning)

### PR DESCRIPTION
**Pull request by hidehalo**
### What problem does this PR solve?

Issue Number: close #16883

Problem Summary:

### What is changed and how it works?

- Show disk usage of a query in slow query log and SLOW_QUERY table
- Show disk usage of a query in statement summary table

What's Changed:

- sessionctx: write disk usage to slow query log item
- infoschema: add disk usage field to statement summary relative tables
- executor: show disk usage in slow query log file
- util: show disk usage in statement summary relative tables

How it Works:

This PR will record the `maximum disk usage` of the query when a slow query is found. And write it to the `slow query log` file and `SLOW_QUERY` table,`STATEMENTS_SUMMARY` table and `STATEMENTS_SUMMARY_HISTORY` table, the following will show changes:

`slow_query.log`/`information_schema.slow_query`
---

Before:

> \# Time: 2019-04-28T15:24:04.309074+08:00
> \# Txn_start_ts: 405888132465033227
> \# Query_time: 0.216905
> \# Cop_time: 0.38 Process_time: 0.021 Request_count: 1 Total_keys: 637 Processed_keys: 436
> \# Is_internal: true
> \# Digest: 42a1c8aae6f133e934d4bf0147491709a8812ea05ff8819ec522780fe657b772
> \# Stats: t1:1,t2:2
> \# Cop_proc_avg: 0.1 Cop_proc_p90: 0.2 Cop_proc_max: 0.03 Cop_proc_addr: 127.0.0.1:20160
> \# Cop_wait_avg: 0.05 Cop_wait_p90: 0.6 Cop_wait_max: 0.8 Cop_wait_addr: 0.0.0.0:20160
> \# Mem_max: 70724
> \# Plan_from_cache: true
> \# Succ: false
> \# Plan_digest: 60e9378c746d9a2be1c791047e008967cf252eb6de9167ad3aa6098fa2d523f4
> \# Prev_stmt: update t set i = 1;
> select * from t;

After:

> \# Time: 2019-04-28T15:24:04.309074+08:00
> \# Txn_start_ts: 405888132465033227
> \# Query_time: 0.216905
> \# Cop_time: 0.38 Process_time: 0.021 Request_count: 1 Total_keys: 637 Processed_keys: 436
> \# Is_internal: true
> \# Digest: 42a1c8aae6f133e934d4bf0147491709a8812ea05ff8819ec522780fe657b772
> \# Stats: t1:1,t2:2
> \# Cop_proc_avg: 0.1 Cop_proc_p90: 0.2 Cop_proc_max: 0.03 Cop_proc_addr: 127.0.0.1:20160
> \# Cop_wait_avg: 0.05 Cop_wait_p90: 0.6 Cop_wait_max: 0.8 Cop_wait_addr: 0.0.0.0:20160
> \# Mem_max: 70724
> \# Disk_max: 65536
> \# Plan_from_cache: true
> \# Succ: false
> \# Plan_digest: 60e9378c746d9a2be1c791047e008967cf252eb6de9167ad3aa6098fa2d523f4
> \# Prev_stmt: update t set i = 1;
> select * from t;

`performance_schema.events_statements_summary_by_digest`/`performance_schema.events_statements_summary_by_digest_history`
---

Before:

> SUMMARY_BEGIN_TIME: 2020-01-02 11:00:00
> SUMMARY_END_TIME: 2020-01-02 11:30:00
> STMT_TYPE: select
> SCHEMA_NAME: test
> DIGEST: 0611cc2fe792f8c146cc97d39b31d9562014cf15f8d41f23a4938ca341f54182
> DIGEST_TEXT: select * from employee where id = ?
> TABLE_NAMES: test.employee
> INDEX_NAMES: NULL
> SAMPLE_USER: root
> EXEC_COUNT: 3
> SUM_LATENCY: 1035161
> MAX_LATENCY: 399594
> MIN_LATENCY: 301353
> AVG_LATENCY: 345053
> AVG_PARSE_LATENCY: 57000
> MAX_PARSE_LATENCY: 57000
> AVG_COMPILE_LATENCY: 175458
> MAX_COMPILE_LATENCY: 175458
> ...........
> AVG_MEM: 103
> MAX_MEM: 103
> AVG_AFFECTED_ROWS: 0
> FIRST_SEEN: 2020-01-02 11:12:54
> LAST_SEEN: 2020-01-02 11:25:24
> QUERY_SAMPLE_TEXT: select * from employee where id=3100
> PREV_SAMPLE_TEXT:
> PLAN_DIGEST: f415b8d52640b535b9b12a9c148a8630d2c6d59e419aad29397842e32e8e5de3
> PLAN:  Point_Get_1     root    1       table:employee, handle:3100

After:

> SUMMARY_BEGIN_TIME: 2020-01-02 11:00:00
> SUMMARY_END_TIME: 2020-01-02 11:30:00
> STMT_TYPE: select
> SCHEMA_NAME: test
> DIGEST: 0611cc2fe792f8c146cc97d39b31d9562014cf15f8d41f23a4938ca341f54182
> DIGEST_TEXT: select * from employee where id = ?
> TABLE_NAMES: test.employee
> INDEX_NAMES: NULL
> SAMPLE_USER: root
> EXEC_COUNT: 3
> SUM_LATENCY: 1035161
> MAX_LATENCY: 399594
> MIN_LATENCY: 301353
> AVG_LATENCY: 345053
> AVG_PARSE_LATENCY: 57000
> MAX_PARSE_LATENCY: 57000
> AVG_COMPILE_LATENCY: 175458
> MAX_COMPILE_LATENCY: 175458
> ...........
> AVG_MEM: 103
> MAX_MEM: 103
> MAX_DISK: 103
> AVG_AFFECTED_ROWS: 0
> FIRST_SEEN: 2020-01-02 11:12:54
> LAST_SEEN: 2020-01-02 11:25:24
> QUERY_SAMPLE_TEXT: select * from employee where id=3100
> PREV_SAMPLE_TEXT:
> PLAN_DIGEST: f415b8d52640b535b9b12a9c148a8630d2c6d59e419aad29397842e32e8e5de3
> PLAN:  Point_Get_1     root    1       table:employee, handle:3100

Tests

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note

UCP: Show disk usage of a query in slow query and statement summary

### CODE REVIEW

**Comment by SunRunAway** [comment]
And it's better to show some output examples of statement summary and slow query log in the `How it Works` section in the PR description.

**Comment by hidehalo** [comment]
> Hi, @hidehalo
> I think we should audit the maximum disk usage, the same as the `memMax` entry in slow query and statement summary.

@SunRunAway Thx for your time, PR is changed

File: executor/adapter.go [hunk 1]
```diff
@@ -829,6 +829,7 @@ func (a *ExecStmt) LogSlowQuery(txnTS uint64, succ bool, hasMoreResults bool) {
 	copTaskInfo := sessVars.StmtCtx.CopTasksDetails()
 	statsInfos := plannercore.GetStatsInfo(a.Plan)
 	memMax := sessVars.StmtCtx.MemTracker.MaxConsumed()
+	diskUsage := sessVars.StmtCtx.DiskTracker.MaxConsumed()
```
Comment (SunRunAway) [inline executor/adapter.go:832]:
```suggestion
	maxDisk := sessVars.StmtCtx.DiskTracker.MaxConsumed()
```

File: infoschema/tables.go [hunk 2]
```diff
@@ -1095,6 +1096,7 @@ var tableStatementsSummaryCols = []columnInfo{
 	{name: "BACKOFF_TYPES", tp: mysql.TypeVarchar, size: 1024, comment: "Types of errors and the number of retries for each type"},
 	{name: "AVG_MEM", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Average memory(byte) used"},
 	{name: "MAX_MEM", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Max memory(byte) used"},
+	{name: "DISK_USAGE", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Disk space(byte) used"},
```
Comment (SunRunAway) [inline infoschema/tables.go:1099]:
```suggestion
	{name: "MAX_DISK", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Max disk space(byte) used"},
```

Comment (hidehalo) [reply infoschema/tables.go:1099]: Make sense.

File: util/stmtsummary/statement_summary.go [hunk 3]
```diff
@@ -812,6 +815,7 @@ func (ssElement *stmtSummaryByDigestElement) toDatum(ssbd *stmtSummaryByDigest)
 		formatBackoffTypes(ssElement.backoffTypes),
 		avgInt(ssElement.sumMem, ssElement.execCount),
 		ssElement.maxMem,
+		ssElement.sumDisk,
```
Comment (djshow832) [inline util/stmtsummary/statement_summary.go:818]: It's better to show the max disk usage and average disk usage.

Comment (hidehalo) [reply util/stmtsummary/statement_summary.go:818]: Yes, it is better.

Comment (hidehalo) [reply util/stmtsummary/statement_summary.go]: @djshow832 Please take a look, I do not understand the reason of CI failed...

File: infoschema/tables.go [hunk 4]
```diff
@@ -1095,6 +1096,7 @@ var tableStatementsSummaryCols = []columnInfo{
 	{name: "BACKOFF_TYPES", tp: mysql.TypeVarchar, size: 1024, comment: "Types of errors and the number of retries for each type"},
 	{name: "AVG_MEM", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Average memory(byte) used"},
 	{name: "MAX_MEM", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Max memory(byte) used"},
+	{name: "MAX_DISK", tp: mysql.TypeLonglong, size: 20, flag: mysql.NotNullFlag | mysql.UnsignedFlag, comment: "Max disk space(byte) used"},
```
Comment (djshow832) [inline infoschema/tables.go:1099]: There should be 2 fields here.

Comment (hidehalo) [reply infoschema/tables.go:1100]: @djshow832 Thank you!

**Comment by you06** [comment]
@hidehalo The points have been fixed, and you have awarded a badge for usability challenge program~ Please fill the form to get your reward! http://tidbcommunity.mikecrm.com/QMCv4QL

**Pull request merged**